dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/ExecutionGraph.java

```diff
@@ -0,0 +1,318 @@
+
+    private Constraint getConstraintFromAxiom(Axiom axiom) {
+        if (constraintMap.containsKey(axiom)) {
+            return constraintMap.get(axiom);
+        }
+
+        Constraint constraint;
+        EventGraph innerGraph = getGraphFromRelation(axiom.getRelation());
+        if (axiom.isAcyclicity()) {
+            constraint = new AcyclicityConstraint(innerGraph);
+        } else if (axiom.isEmptiness()) {
+            constraint = new EmptinessConstraint(innerGraph);
+        } else if (axiom.isIrreflexivity()) {
+            constraint = new IrreflexivityConstraint(innerGraph);
+        } else {
+            throw new UnsupportedOperationException("The axiom " + axiom + " is not recognized.");
+        }
+
+        constraintMap.put(axiom, constraint);
+        return constraint;
+    }
+
+    private EventGraph getGraphFromRelation(Relation rel) {
+        if (relationGraphMap.containsKey(rel)) {
+            return relationGraphMap.get(rel);
+        }
+
+        EventGraph graph;
+        Class<?> relClass = rel.getClass();
+
+        // ===== Filter special relations ======
+        if (SPECIAL_RELS.contains(rel.getName())) {
+            switch (rel.getName()) {
+                case CTRL:
+                    graph = new CtrlDepGraph();
+                    break;
+                case DATA:
+                    graph = new DataDepGraph();
+                    break;
+                case ADDR:
+                    graph = new AddrDepGraph();
+                    break;
+                case CRIT:
+                    graph = new RcuGraph();
+                    break;
+                default:
+                    throw new UnsupportedOperationException(rel.getName() + " is marked as special relation but has associated graph.");
+            }
+        } else if (relClass == RelRf.class) {
+            graph = rfGraph = new ReadFromGraph();
+        } else if (relClass == RelLoc.class) {
+            graph = locGraph = new LocationGraph();
+        } else if (relClass == RelPo.class) {
+            graph = poGraph = new ProgramOrderGraph();
+        } else if (relClass == RelCo.class) {
+            graph = coGraph = new TransitiveGraph(woGraph = new CoherenceGraph());
+            woGraph.setName("_wo");
+        } else if (rel.isRecursiveRelation()) {
+            RecursiveGraph recGraph = new RecursiveGraph();
+            recGraph.setName(rel.getName() + "_rec");
+            relationGraphMap.put(rel, recGraph);
+            recGraph.setConcreteGraph(getGraphFromRelation(rel.getInner()));
+            return recGraph;
+        } else if (rel.isUnaryRelation()) {
+            Relation innerRelation = rel.getInner();
+            EventGraph innerGraph = getGraphFromRelation(innerRelation);
+            // A safety check because recursion might compute this edge set
+            if (relationGraphMap.containsKey(rel)) {
+                return relationGraphMap.get(rel);
```

Comment:
They are not useless. As the comment suggests, due to recursive definitions it might happen that the call to ```getGraphFromRelation``` to the inner graphs might end up creating the current graph already. I'm not sure if this ever happens to be honest, but that's why it is a "safety check" I guess. But I will take a look into the cases of when this code actually gets executed.

Reply:
Sorry, I guess the comment was not clear. I don't say the check is useless, but the two lines BEFORE the check are useless if the check succeeds and thus I would move the check before those lines.

Reply:
That's what I'm saying, those two lines BEFORE the check actually DO influence the check cause they potentially update the ```relationGraphMap```. And in the case of recursive definitions, these calls may already have constructed the graph that the code is about to construct (so it aborts and returns the already constructed graph). Hence, moving the check before the calls to ```getGraphFromRelation``` will make the check meaningless.

Reply:
ok, now I get it. I'll mark this and the similar one as resolved.